hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RouterObserverReadProxyProvider.java

```diff
@@ -84,7 +84,8 @@ public class RouterObserverReadProxyProvider<T> extends AbstractNNFailoverProxyP
 
   public RouterObserverReadProxyProvider(Configuration conf, URI uri, Class<T> xface,
       HAProxyFactory<T> factory) {
-    this(conf, uri, xface, factory, new IPFailoverProxyProvider<>(conf, uri, xface, factory));
+    this(conf, uri, xface, factory,
+        new ConfiguredFailoverProxyProvider<>(conf, uri, xface, factory));
```

Comment:
In my original code I assumed the routers would all sit behind a single host name and traffic would be split via DNS. The IpFailoverProxyProvider is then needed incase one router is down but the IP is still in the DNS record.

I can see how in other setups users may need to list all the routers explicitly in the client configurations.

Can we make the proxy provider configurable? To support both usecases.

Reply:
Actually, `IPFailoverProxyProvider` can still be set on `routerContext.getFileSystemURI().toString()` even when we use the ConfiguredFailoverProxyProvider.

Reply:
What do you think about making it configurable? For a different internal proxyprovider we have code like below.
```
this(conf, uri, xface, proxyFactory, getFailoverProxyProvider(conf, uri, xface, proxyFactory));
```
and
```
  @VisibleForTesting
  protected static <T> AbstractNNFailoverProxyProvider<T> getFailoverProxyProvider(
      Configuration conf, URI uri, Class<T> xface, HAProxyFactory<T> proxyFactory) {
    String configKey = FAILOVER_PROXY_PROVIDER + "." + uri.getHost();
    Class<AbstractNNFailoverProxyProvider<T>> fppClass =
        (Class<AbstractNNFailoverProxyProvider<T>>) conf.getClass(configKey,
            (Class<? extends AbstractNNFailoverProxyProvider>) DEFAULT_FAILOVER_PROXY_PROVIDER,
            AbstractNNFailoverProxyProvider.class);

    try {
      Constructor<AbstractNNFailoverProxyProvider<T>> ctor = fppClass.getConstructor(
          Configuration.class, URI.class, Class.class, HAProxyFactory.class);
      return ctor.newInstance(conf, uri, xface, proxyFactory);
    } catch (Exception e) {
      LOG.error("Failed to create failover proxy provider", e);
      throw new IllegalArgumentException(e);
    }
  }
  ```

Reply:
@simbadzina  Thanks for your review and sharing your thoughts. I understand that you intentionally chose `IPFailoverProxyProvider`.

> Can we make the proxy provider configurable? To support both usecases.

Are you suggesting that we create another configuration like `dfs.client.failover.router.internal.proxy.provider`?

Reply:
As an alternative approach, what if we created a `RouterObserverReadProxyProviderWithIPFailover` using `IPFailoverProxyProvider` to align it with `ObserverReadProxyProviderWithIPFailover`?

```java
public class RouterObserverReadProxyProviderWithIPFailover<T>  
    extends RouterObserverReadProxyProvider<T> {  
  
  @VisibleForTesting  
  static final Logger LOG =  
      LoggerFactory.getLogger(RouterObserverReadProxyProviderWithIPFailover.class);  
  
  public RouterObserverReadProxyProviderWithIPFailover(Configuration conf, URI uri, Class<T> xface,  
      HAProxyFactory<T> factory) {  
    super(conf, uri, xface, factory, new IPFailoverProxyProvider<>(conf, uri, xface, factory));  
  }  
}
```

Reply:
Yes, I was suggesting adding a new parameter, like `dfs.client.failover.router.internal.proxy.provider` you named.

Creating a new class is also a good solution. I'm a bit worried though about the update story for clients who are already using the existing class.

The new parameter approach makes the update backward compatible with existing client configs.

Reply:
I'd prefer to create a new class. I'd like to keep the existing class as is and create a new class named `RouterObserverReadConfiguredFailoverProxyProvider` using `ConfiguredFailoverProxyProvider` internally. I will update the PR soon.

Reply:
Cool. That sounds good to me.